packages/devtools_app/lib/src/screens/debugger/codeview.dart

```diff
@@ -295,127 +318,137 @@ class _CodeViewState extends State<CodeView>
-                                child: SingleChildScrollView(
-                                  scrollDirection: Axis.horizontal,
-                                  controller: horizontalController,
-                                  child: SizedBox(
-                                    height: constraints.maxHeight,
-                                    width: math.max(
-                                      constraints.maxWidth,
-                                      fileWidth,
-                                    ),
-                                    child: Lines(
-                                      height: constraints.maxHeight,
-                                      debugController: widget.controller,
-                                      scrollController: textController,
-                                      lines: lines,
-                                      pausedFrame: pausedFrame,
-                                      searchMatchesNotifier:
-                                          widget.controller.searchMatches,
-                                      activeSearchMatchNotifier:
-                                          widget.controller.activeSearchMatch,
-                                    ),
-                                  ),
+    final contentBuilder = (context, ScriptRef? script) {
+      if (lines.isNotEmpty) {
+        return DefaultTextStyle(
+          style: theme.fixedFontStyle,
+          child: Scrollbar(
+            key: CodeView.debuggerCodeViewVerticalScrollbarKey,
+            controller: textController,
+            thumbVisibility: true,
+            // Only listen for vertical scroll notifications (ignore those
+            // from the nested horizontal SingleChildScrollView):
+            notificationPredicate: (ScrollNotification notification) =>
+                notification.depth == 1,
+            child: ValueListenableBuilder<StackFrameAndSourcePosition?>(
+              valueListenable: widget.debuggerController?.selectedStackFrame ??
+                  const FixedValueListenable<StackFrameAndSourcePosition?>(
+                    null,
+                  ),
+              builder: (context, frame, _) {
+                final pausedFrame = frame == null
+                    ? null
+                    : (frame.scriptRef == scriptRef ? frame : null);
+
+                return Row(
+                  children: [
+                    ValueListenableBuilder<List<BreakpointAndSourcePosition>>(
+                      valueListenable:
+                          breakpointManager.breakpointsWithLocation,
+                      builder: (context, breakpoints, _) {
+                        return Gutter(
+                          gutterWidth: gutterWidth,
+                          scrollController: gutterController,
+                          lineCount: widget.lineRange?.size ?? lines.length,
+                          lineOffset: (widget.lineRange?.begin ?? 1) - 1,
```

Comment:
what if widget.lineRange.begin = 0?

Reply:
This shouldn't happen as line numbers always start at 1.